### PR TITLE
441 fine tuning intercepting endpoints

### PR DESCRIPTION
**Pull request by Intex32**
sub PR of #464 
cloned OAI endpoints for fine tuning (to later intercept and collect metrics)

### CODE REVIEW

File: server/build.gradle.kts [hunk 1]
```diff
@@ -57,6 +58,8 @@ dependencies {
     implementation(projects.xefCore)
     implementation(projects.xefLucene)
     implementation(projects.xefPostgresql)
+    implementation("io.ktor:ktor-server-core-jvm:2.3.4")
```
Comment (raulraja) [inline server/build.gradle.kts:61]: Dependencies should be added in toml file

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import io.ktor.http.content.*
+import io.ktor.server.application.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.util.*
+import  io.ktor.http.URLBuilder
+import io.ktor.util.pipeline.*
+import io.ktor.utils.io.jvm.javaio.*
+
+private const val OAI_URL = "https://api.openai.com"
```
Comment (raulraja) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt:18]: In other endpoints, like chatCompletions, we are also accessing this URL. Is this a newly introduced constant? Can we pull this value automatically from the OpenAi lib deps?

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.xebia.functional.xef.server.http.routes.providers
+
+import com.xebia.functional.xef.server.http.routes.providers.openai.oaiFiles
+import com.xebia.functional.xef.server.http.routes.providers.openai.oaiFineTuning
+import io.ktor.client.*
+import io.ktor.server.auth.*
+import io.ktor.server.routing.*
+
+fun Routing.oaiRoutes(
```
Comment (raulraja) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt:9]: If this is the `oaiRoutes`, wouldn't chat completion and embeddings be part of these? I think we have some confusion regarding how routes are organized in the Xef server.

**Review by raulraja — CHANGES_REQUESTED** [review]

File: server/build.gradle.kts (hunk 1 above)
Comment (Intex32) [reply server/build.gradle.kts:61]: Idk honestly know where this time comes from. We have these dependencies already from the toml config. I'll remove it.

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt (hunk 2 above)
Comment (Intex32) [reply server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt:18]: The constant used by chatCompletions additionally holds the path /v1/ and is a hardcode in the endpoint's file.
This url should only contain the base url, as the path and parameters are copied automatically from the incoming request with the new utility functions I introduced in this PR.

It is not desirable to use this variable from the Aallam library, as the server shouldn't have the library as a dependency right. We could create a variable in xef-core like DEFAULT_OPENAI_URL that can be used by the xef-server. Furthermore, I just looked it up, the Aallam's url to OpenAI contains the v1 path too.

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt (hunk 3 above)
Comment (Intex32) [reply server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt:9]: Yes, @Montagon had to curb me, not too change too much in one PR again. The other endpoints, if approved, will have to get integrated with the my new structure. This includes the forwarding logic and registering the endpoints as here.

Comment (Montagon) [reply server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt:9]: I said this comment to you referred to when you have the other PR in this one.

Comment (Intex32) [reply server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/Routes.kt:9]: The same applies here, no? You made some changes in the existing endpoints and files. I didn't want to mess (or get in conflict) with that. You instructed me to not change the existing endpoints at that point, me acuerdo.

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/XefRoutes.kt [hunk 4]
```diff
@@ -13,5 +12,4 @@ fun Routing.xefRoutes(
   userRoutes(UserRepositoryService(logger))
   organizationRoutes(OrganizationRepositoryService(logger))
   projectsRoutes(ProjectRepositoryService(logger))
-  tokensRoutes(TokenRepositoryService(logger))
```
Comment (javipacheco) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/XefRoutes.kt:16]: Why are we removing these routes?

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+ * As we are copying the API from the providers,
+ * everything including the path structure has follow the API contract.
+ */
+private fun buildProviderUrlFromRequest(request: ApplicationRequest) =
+    URLBuilder().takeFrom(OAI_URL).apply {
+        encodedPath = request.path()
+    }.build()
+
+@Deprecated("")
+internal suspend fun HttpClient.makeRequest(
```
Comment (javipacheco) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt:32]: We are not using this method  and I don't understand why is deprecated without a description

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        }
+        contentType(ContentType.Application.Json)
+        method = HttpMethod.Post
+        setBody(body)
+    }
+    call.response.headers.copyFrom(response.headers)
+    call.respond(response.status, response.body<String>())
+}
+
+@Deprecated("")
```
Comment (javipacheco) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/providers/RequestForwardingUtil.kt:50]: Same above

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/AIRoutes.kt [hunk 7]
```diff
@@ -62,6 +62,7 @@ fun Routing.aiRoutes(
     }
 }
 
+@Deprecated("will be removed in future pr")
```
Comment (javipacheco) [inline server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/AIRoutes.kt:65]: We don't have any issue for that.  For me, this is not a valid description for a deprecated function

**Review by javipacheco — CHANGES_REQUESTED** [review]

**Comment by raulraja** [comment]
We are going to put the content of this PR on hold because we are still evaluating a different approach for the server https://github.com/xebia-functional/xef/issues/362#issuecomment-1740714316 and re-considering server development priorities based on how it's getting used in our current clients. We will revisit the fine-tuning endpoints and decide whether to include this PR or not after we gather feedback on how fine tuning will be used in the client and if we need anything beyond what the OAI interface provides.

File: server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/XefRoutes.kt (hunk 4 above)
Comment (Intex32) [reply server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/XefRoutes.kt:16]: I had some struggle splitting the PR in two. I was in merge hell rebasing and cherrypicking. I tried to mitigate human mistakes. That one's on me.

**Comment by tomascayuelas** [comment]
Can we close this Draft @raulraja @Montagon @javipacheco ?

**Pull request closed without merging**